src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java

```diff
@@ -0,0 +1,690 @@
+            .of()
+            .sku("sku")
+            .key("new-product-master-variant")
+            .attributes(newCustomObjectReferenceAttribute)
+            .build();
+
+        final ProductDraft newProductDraftWithCustomObjectReference = ProductDraftBuilder
+            .of(productType, ofEnglish("productName"), ofEnglish("productSlug"), newMasterVariant)
+            .key("new-product")
+            .build();
+
+
+        // test
+        final ProductSync productSync = new ProductSync(syncOptions);
+        final ProductSyncStatistics syncStatistics =
+            productSync
+                .sync(singletonList(newProductDraftWithCustomObjectReference))
+                .toCompletableFuture()
+                .join();
+
+        // assertion
+        assertThat(syncStatistics).hasValues(1, 0, 1, 0, 0);
+        assertThat(errorCallBackExceptions).isEmpty();
+        assertThat(errorCallBackMessages).isEmpty();
+        assertThat(warningCallBackMessages).isEmpty();
+
+        final ObjectNode expectedNestedAttributeValue =
+            createNestedAttributeValueReferences(CUSTOM_OBJECT_REFERENCE_ATTR_NAME,
+                createReferenceObject(testCustomObject2.getId(), CustomObject.referenceTypeId()));
+
+        final AttributeDraft expectedCustomObjectReferenceAttribute =
+            AttributeDraft.of("nestedAttribute", createArrayNode(expectedNestedAttributeValue));
+        assertThat(actions).containsExactly(SetAttribute.of(1, expectedCustomObjectReferenceAttribute, true));
+
+
+        final Product createdProduct = CTP_TARGET_CLIENT
+            .execute(ProductByKeyGet.of(productDraftWithCustomObjectReference.getKey()))
+            .toCompletableFuture()
+            .join();
+
+        final Optional<Attribute> createdCustomObjectReferenceAttribute =
+            createdProduct.getMasterData().getStaged().getMasterVariant()
+                          .findAttribute(customObjectReferenceAttribute.getName());
+
+        assertThat(createdCustomObjectReferenceAttribute).hasValueSatisfying(attribute -> {
+            final JsonNode nestedAttributeNameField = attribute
+                .getValueAsJsonNode()
+                .get(0)
+                .get(ATTRIBUTE_NAME_FIELD);
+            final JsonNode nestedAttributeValueField = attribute
+                .getValueAsJsonNode()
+                .get(0)
+                .get(ATTRIBUTE_VALUE_FIELD);
+
+
+            assertThat(nestedAttributeNameField.asText()).isEqualTo(CUSTOM_OBJECT_REFERENCE_ATTR_NAME);
+            assertThat(nestedAttributeValueField.get(REFERENCE_TYPE_ID_FIELD).asText())
+                .isEqualTo(CustomObject.referenceTypeId());
+            assertThat(nestedAttributeValueField.get(REFERENCE_ID_FIELD).asText()).isEqualTo(testCustomObject2.getId());
+        });
+    }
+
+    @Test
+    void sync_withNonExistingNestedCustomObjectReferenceAsAttribute_ShouldFailCreatingTheProduct() {
```

Comment:
So it means, we expect in _project-sync_, we execute `CustomObjectSyncer` before we execute `ProductSyncer`?

Reply:
yes, I haven't' check the custom object part on the project sync, I assume @lojzatran is considering it.

Reply:
I do it like that in the project-sync (custom object before product), but this actually should not matter as they're not referenced to each other.